Exposing some hidden CFunctionReflection members to the front end (ID3D12LibraryReflection/ID3D12FunctionReflection)

Nielsbishere opened the pull request:
The DXC backend has the following parameter:
```const DxilFunctionProps *m_pProps```
Which has a large number of members that are not accessible by ID3D12FunctionReflection and D3D12_FUNCTION_DESC.
The most interesting ones being raytracing and workgraph related.
Exposing these saves me a lot of headache, as I won't need to parse these manually, especially since they're already available in DXC.

The following parameters still need to be exposed (and I'll take a look later to add them):
- Serialized root signature could have a getter which returns a DxcBuffer.
- InputNodes / OutputNodes could have a getter which returns a CNodeIOProperties* and length. Where each CNodeIOProperties has const char* for name as well as UINT id.
- NodeID NodeShaderId, NodeShaderSharedInput could use UINT ids in the D3D12_NODE_SHADER_DESC struct, while leaving the strings as getters.

The following parameters I won't add, since I have no idea how to properly expose them and if they're even useful to the front end:
- Vertex reflection: clipPlanes[6] returns an llvm::Constant
- Hull reflection: returns patchConstFunction as llvm::Function, which won't be accessible.
Since these are llvm internal values, I'm guessing that they might not be very useful for the front end.

This relates to a DXCompiler issue & PR.

jenatali commented:
> This relates to a DXCompiler issue & PR.

Please link the actual PR?

I don't think we can merge a *change* to the `D3D12_SHADER_DESC` structure, since that structure is returned from both DXC and FXC, and FXC has versioning constraints that can't break. I think we can add new structures to this header, assuming the DXC changes are reviewed and accepted.

jenatali reviewed (changes requested):
I don't think we can merge a change to the D3D12_SHADER_DESC structure, since that structure is returned from both DXC and FXC, and FXC has versioning constraints that can't break. I think we can add new structures to this header, assuming the DXC changes are reviewed and accepted.

Nielsbishere commented:
Okay sounds good, I'll add an extended version that can only be queried for DXC. The DXC version is on my fork, but I gotta rebase it first, since it's built on other changes.

Nielsbishere commented:
Implemented by https://github.com/microsoft/DirectXShaderCompiler/pull/6827 on the DXC side. Not sure if there's anything that has to be done for FXC? Maybe it shouldn't be declared "PURE" and return failed instead?

jenatali commented:
> Implemented by [microsoft/DirectXShaderCompiler#6827](https://github.com/microsoft/DirectXShaderCompiler/pull/6827) on the DXC side. Not sure if there's anything that has to be done for FXC? Maybe it shouldn't be declared "PURE" and return failed instead?

The way to do this without breaking FXC is to put it on an `ID3D12FunctionReflection1` with a new IID, and then you can implement `QueryInterface` for it.

Specifically, the layout of the virtual function table cannot change. Inserting a function in the middle of the table means that when you call that function, or later, on a concrete implementation coming from FXC or old DXC, you're calling a different function than you think you are. You could add it to the end, but then if you try to call it on an old concrete implementation, you'll just crash because there's *nothing* there in the function table. Even if you have a default implementation instead of `PURE`, you'd need to recompile the caller, for it to have a function at that spot in the vtable.

The only way to have a clean runtime failure is to add a separate interface and query.

Nielsbishere commented:
It seems like I have to expose this with ID3D12LibraryReflection1::GetFunctionByIndex1, as ID3D12FunctionReflection doesn't inherrit IUnknown, which I don't think I can safely change. I can make the new ID3D12FunctionReflection1 inherit IUnknown for the future though... But not ideal any way.

damyanp commented on include/directx/d3d12shader.h:
```diff
@@ -119,7 +119,7 @@ typedef struct _D3D12_SHADER_DESC
     LPCSTR                  Creator;                     // Creator string
     UINT                    Flags;                       // Shader compilation/parse flags
-    
+
```
Please remove unrelated changes.

```suggestion

```

damyanp commented on include/directx/d3d12shader.h:
```diff
@@ -208,54 +260,95 @@ typedef struct _D3D12_SHADER_INPUT_BIND_DESC
 #define D3D_SHADER_REQUIRES_SAMPLE_CMP_GRADIENT_OR_BIAS                                     0x80000000
 #define D3D_SHADER_REQUIRES_EXTENDED_COMMAND_INFO                                           0x100000000ull
 
-
```
Please revert changes unrelated to this change.

damyanp commented on include/directx/d3d12shader.h:
```diff
@@ -208,54 +260,95 @@ typedef struct _D3D12_SHADER_INPUT_BIND_DESC
-    
-    UINT                    ConstantBuffers;             // Number of constant buffers
-    UINT                    BoundResources;              // Number of bound resources
+    UINT                                    Version;                      // Shader version
```
It looks like maybe there's no actual changes to D3D12_FUNCTION_DESC?  If that's the case, please revert these changes so that we can focus on reviewing the actual changes proposed.

damyanp reviewed (commented):
There's quite a bit of API surface here that needs more careful review. So far, I've only looked at things that could be done to make the review process easier.

Nielsbishere commented:
Implemented PR changes. Also exposed all remaining properties such as node info and serialized root signature.
Only thing left to do could be implementing this in D3D12_SHADER_DESC1 as well. But that could be a PR for later.

tex3d commented on include/directx/d3d12shader.h:
```diff
@@ -208,54 +260,95 @@ typedef struct _D3D12_SHADER_INPUT_BIND_DESC
     UINT      FunctionCount;     // Number of functions exported from the library.
 } D3D12_LIBRARY_DESC;
 
+typedef enum D3D12_NODE_OVERRIDES_TYPE {
```
`D3D12_NODE_OVERRIDES_TYPE` is a D3D12 enum for specifying which parameter overrides are specified in a union passed to the API when constructing a graph.  It may look like an enum that defines launch type, but that's not its purpose.

For example, `D3D12_NODE_OVERRIDES_TYPE_COMMON_COMPUTE` has no corresponding launch type, instead indicating that overrides common to any launch type are being specified.  Imagine a new hypothetical launch type was added - just for fun let's call it 'mesh' and assign it a value of 4.  That would collide with the COMMON_COMPUTE enum value.

`D3D12_NODE_SHADER_DESC` should use an enum specifically defining launch type, like `D3D12_NODE_LAUNCH_TYPE`, rather than reuse the `D3D12_NODE_OVERRIDES_TYPE` enum.

Nielsbishere replied:
Done. The only thing I'm not 100% sure about is the DXC backend. At the time of CFunctionReflection being available, is D3D12_NODE_OVERRIDES_TYPE_COMMON_COMPUTE  already resolved or do I need to explicitly catch it and turn it into a different D3D12_NODE_LAUNCH_TYPE?

Nielsbishere replied:
Okay good, should be good to go then :)

jenatali reviewed (approved):
I'll let the DXC folks drive whether the corresponding implementation PR in the compiler repo is appropriate, but looking solely at the API surface exposed here, this seems acceptable to me.

damyanp commented:
See https://github.com/microsoft/DirectXShaderCompiler/pull/6827#issuecomment-2260908647

We need to make sure we've completed a thorough review of all of this new API surface that'll need to be supported for a long time before accepting the change to the headers.

Thanks!

jenatali commented:
Yes, agreed, I have no intentions of actually merging this until the corresponding DXC change is reviewed and completed. Assuming that the implementation is accepted, we can take this into the headers and the Windows SDK afterwards.

jenatali commented on include/directx/d3d12shader.h:
```diff
@@ -416,6 +608,10 @@ DECLARE_INTERFACE_(ID3D12ShaderReflection, IUnknown)
 
     STDMETHOD_(UINT64, GetRequiresFlags)(THIS) PURE;
+
+    STDMETHOD_(BOOL, GetWaveSize)
```
This kind of breaking change made sense with the FXC model, where you linked against and shipped a redist `D3DCompiler_xx.dll`, and also makes sense with a strictly-DXC model. But in a world where `D3DCompiler_47.dll` is inbox in Windows and implements the old interface, where DXC can implement a new one, I think I'd prefer to rev this in a non-breaking way.

We can do that by just adding `ID3D12ShaderReflection1`, which inherits from `ID3D12ShaderReflection` and adds the new method. The VTable of `ID3D12ShaderReflection1` is then identical to this new definition, except that apps can still query for the older one to avoid it being a source-breaking change for FXC consumers.

Nielsbishere commented:
Since last time, added numthreads to mesh/amplification shaders and implemented ID3D12ShaderReflection1.
There were some... issues with upgrading DXC to the latest DirectX-Headers, since it was behind by 2 years.

**Important notes:** DXC uses the assumption that bool* == LPBOOL, which isn't possible if BOOL is uint32_t, this needs some further attention to see if this is OK or has side effects, but DXC and DirectX12-Headers definitely don't agree on the size of a bool... Checking the define for LLVM_SUPPORT_WIN_ADAPTER_H is used to avoid duplicate defines and structs/typedefs, this is required on Linux since basestd.h from d3dcommon.h has its own definitions that are definitely different than from DXC (dxc/WinAdapter.h).

Nielsbishere commented:
This branch is based of #137 to allow DXC to update to latest DirectX-Headers. DirectX-Headers itself should be fine now, just need to add support for the reflection data to dxcdisassembler and update unit tests to verify dxa output.

Nielsbishere commented:
Added an important function to ID3D12FunctionReflection1; which is GetRequiresFlags such as in ID3D12ShaderReflection. Without this, it is impossible to find out what features are being used by a lib file (I thought the desc Flags was used for this, but nope).

This is especially important now that some GPUs don't always support the base featureset, so I use the following to check feature support of the underlying GPUs:
```
		D3D_SHADER_REQUIRES_RAYTRACING_TIER_1_1,
		D3D_SHADER_REQUIRES_NATIVE_16BIT_OPS,
		D3D_SHADER_REQUIRES_INT64_OPS,
		D3D_SHADER_REQUIRES_VIEW_ID,
		D3D_SHADER_REQUIRES_DOUBLES,
		D3D_SHADER_REQUIRES_11_1_DOUBLE_EXTENSIONS,
		D3D_SHADER_REQUIRES_ATOMIC_INT64_ON_TYPED_RESOURCE,
		D3D_SHADER_REQUIRES_ATOMIC_INT64_ON_GROUP_SHARED,
		D3D_SHADER_REQUIRES_DERIVATIVES_IN_MESH_AND_AMPLIFICATION_SHADERS,
		D3D_SHADER_REQUIRES_WRITEABLE_MSAA_TEXTURES,
		D3D_SHADER_REQUIRES_WAVE_OPS
```

I use this reflection to know whether a shader is really using a feature. It's possible someone enables I64 to include something that requires it but never actually uses I64 in the final shader. This is then used to retroactively disable that feature to ensure we don't go checking features that aren't really relevant.
